### ISSUE TITLE
[no-undef] False positive with default argument in a class method

### ISSUE THREAD

**Issue by sk-** [typescript-eslint/typescript-eslint#2956]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "root": true,
  "parserOptions": {
    "ecmaVersion": 2018,
    "project": "./tsconfig.json"
  },
  "env": {
    "node": true,
    "es6": true
  },
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint"],
  "rules": {
    "@typescript-eslint/await-thenable": "error",
    //"@typescript-eslint/no-floating-promises": "error",
    "@typescript-eslint/no-misused-promises": "error",
    //"@typescript-eslint/no-unnecessary-boolean-literal-compare": "error",
    //"@typescript-eslint/no-unnecessary-condition": "error",
    "@typescript-eslint/prefer-includes": "error",
    "@typescript-eslint/prefer-string-starts-ends-with": "error",
    "@typescript-eslint/switch-exhaustiveness-check": "error"
  }
}
```

```js
class A {
    /**
     * @param {Date} [date] If not specified, will use the current time
     * @return {Date}
     */
    test(date = new Date(Date.now())) {
        return date;
    }
}
```

```json
{
  "compilerOptions": {
    "noEmit": true,
    "module": "commonjs",
    "target": "es2019",
    "lib": ["es2020"],
    "allowJs": true,
    "checkJs": true,
    "strict": true,
    "resolveJsonModule": true,
    // Reenable
    "noImplicitAny": false,
  },
  "include": [
    "**/*.js",
    "**/*.ts",
  ]
}
```

**Expected Result**

No `no-undef` error should be thrown.

**Actual Result**

```
functions/test.js
   6:31  error  'now' is not defined               no-undef
```
**Additional Info**

The check `no-undef` only triggers the error in class methods with default arguments. If I move the test function out of the class, then no exception is reported.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.14.0` |
| `@typescript-eslint/parser`        | `4.14.0` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `7.18.0` |
| `node`                             | `14.13.1` |

**Comment by bradzacher** [comment]
Same as #2941, #2942, #2945, #2947, #2950

Please use the search
